File: Anime4K-Upscaler/ViewModels/CompressViewModel.swift
// Anime4K-Upscaler/ViewModels/CompressViewModel.swift
// ViewModel for the Compress feature — file management, configuration, processing engine.

import Foundation
import SwiftUI
import UniformTypeIdentifiers
import Observation
import IOKit.pwr_mgt

/// Central ViewModel for the Compress feature.
/// Manages file list, encoder/quality/content configuration, HDR detection,
/// and FFmpeg process execution with progress tracking.
@MainActor @Observable
final class CompressViewModel {

    // MARK: - File State

    var files: [VideoFile] = []
    var selectedFileID: UUID?

    // MARK: - Configuration

    var encoder: CompressEncoder = .hevcVideoToolbox
    var quality: Int = 68
    var contentType: ContentType = .liveAction
    var bFrames: Int = 3
    var longGOPEnabled: Bool = false

    // MARK: - Output

    var outputDirectoryURL: URL?

    var outputDirectoryDisplayName: String {
        outputDirectoryURL?.lastPathComponent ?? "Not selected"
    }

    // MARK: - Processing State

    var viewState: ViewState = .configuration
    var isProcessing: Bool = false
    var currentJobIndex: Int = 0
    var totalJobs: Int = 0
    var overallProgress: Double = 0.0
    var jobs: [CompressJob] = []

    enum ViewState: String {
        case configuration
        case processing
    }

    // MARK: - Internal Engine State

    @ObservationIgnored private var currentProcess: Process?
    @ObservationIgnored private var powerAssertionID: IOPMAssertionID = 0
    @ObservationIgnored private var hasPowerAssertion: Bool = false
    @ObservationIgnored private var cancellationRequested: Bool = false
    @ObservationIgnored private var activityToken: NSObjectProtocol?

    private static let uiUpdateIntervalMs: Int = 100

    // MARK: - File Management

    func addFiles() {
        let urls = SecurityScopeManager.shared.presentVideoFilePicker(allowMultiple: true)
        guard !urls.isEmpty else { return }

        for url in urls {
            guard !files.contains(where: { $0.url == url }) else { continue }
            let ext = url.pathExtension.lowercased()
            guard SupportedVideoExtension.allExtensions.contains(ext) else { continue }

            var videoFile = VideoFile(url: url)
            videoFile.bookmarkData = SecurityScopeManager.shared.createBookmark(for: url)
            files.append(videoFile)
        }

        probeNewFiles()

        if selectedFileID == nil {
            selectedFileID = files.first?.id
        }
    }

    func addFilesFromDrop(_ urls: [URL]) {
        for url in urls {
            guard !files.contains(where: { $0.url == url }) else { continue }
            let ext = url.pathExtension.lowercased()
            guard SupportedVideoExtension.allExtensions.contains(ext) else { continue }

            SecurityScopeManager.shared.startAccessing(url)
            var videoFile = VideoFile(url: url)
            videoFile.bookmarkData = SecurityScopeManager.shared.createBookmark(for: url)
            files.append(videoFile)
        }

        probeNewFiles()

        if selectedFileID == nil {
            selectedFileID = files.first?.id
        }
    }

    func removeFile(id: UUID) {
        if let index = files.firstIndex(where: { $0.id == id }) {
            SecurityScopeManager.shared.stopAccessing(files[index].url)
            files.remove(at: index)
        }
        if selectedFileID == id {
            selectedFileID = files.first?.id
        }
    }

    func removeAllFiles() {
        for file in files {
            SecurityScopeManager.shared.stopAccessing(file.url)
        }
        files.removeAll()
        selectedFileID = nil
    }

    // MARK: - Metadata Probing

    private func probeNewFiles() {
        let unprobed = files.filter { $0.durationSeconds == nil }
        guard !unprobed.isEmpty else { return }
        let urls = unprobed.map(\.url)

        Task {
            let results = await DurationProbe.batchProbe(urls: urls)
            for (url, result) in results {
                if let index = files.firstIndex(where: { $0.url == url }),
                   let probeResult = result {
                    files[index].durationSeconds = probeResult.durationSeconds
                    files[index].width = probeResult.width
                    files[index].height = probeResult.height
                }
            }
        }
    }

    // MARK: - Configuration Helpers

    func onEncoderChanged() {
        quality = encoder.defaultQuality
    }

    func updateQuality(_ value: Int) {
        quality = max(0, min(value, encoder.maxQuality))
    }

    var selectedFile: VideoFile? {
        guard let id = selectedFileID else { return nil }
        return files.first(where: { $0.id == id })
    }

    var totalFileSize: String {
        let total = files.reduce(Int64(0)) { $0 + $1.fileSizeBytes }
        return ByteCountFormatter.string(fromByteCount: total, countStyle: .file)
    }

    var totalDuration: String? {
        let durations = files.compactMap(\.durationSeconds)
        guard !durations.isEmpty else { return nil }
        let total = durations.reduce(0, +)
        let hours = Int(total) / 3600
        let minutes = (Int(total) % 3600) / 60
        let seconds = Int(total) % 60
        return String(format: "%02d:%02d:%02d", hours, minutes, seconds)
    }

    var canStartProcessing: Bool {
        !files.isEmpty && !isProcessing
    }

    var batchSummary: String {
        "\(files.count) file\(files.count == 1 ? "" : "s") • \(encoder.displayName) • \(contentType.displayName)"
    }

    var completedJobCount: Int {
        jobs.filter { $0.state == .completed }.count
    }

    var failedJobCount: Int {
        jobs.filter { $0.state == .failed }.count
    }

    var allJobsFinished: Bool {
        !jobs.isEmpty && jobs.allSatisfy { $0.state.isTerminal }
    }

    // MARK: - Output Directory

    func selectOutputDirectory() {
        if let url = SecurityScopeManager.shared.presentOutputDirectoryPicker() {
            outputDirectoryURL = url
        }
    }

    // MARK: - Processing Control

    func startProcessing() {
        guard canStartProcessing else { return }

        if outputDirectoryURL == nil {
            selectOutputDirectory()
            guard outputDirectoryURL != nil else { return }
        }

        let config = CompressConfiguration(
            encoder: encoder,
            quality: quality,
            contentType: contentType,
            bFrames: bFrames,
            longGOPEnabled: longGOPEnabled
        )

        jobs = files.map { file in
            CompressJob(file: file, configuration: config, outputDirectory: outputDirectoryURL)
        }

        viewState = .processing

        Task {
            await executeBatch(jobs: jobs)
        }
    }

    func cancelProcessing() {
        cancellationRequested = true
        if let process = currentProcess, process.isRunning {
            process.terminate()
        }
    }

    func returnToConfiguration() {
        guard !isProcessing else { return }
        viewState = .configuration
    }

    // MARK: - Batch Execution

    private func executeBatch(jobs: [CompressJob]) async {
        guard !jobs.isEmpty else { return }

        isProcessing = true
        totalJobs = jobs.count
        currentJobIndex = 0
        cancellationRequested = false

        acquirePowerAssertion()
        beginAppNapPrevention()

        for (index, job) in jobs.enumerated() {
            if cancellationRequested { break }
            currentJobIndex = index + 1
            await executeJob(job)
            overallProgress = Double(index + 1) / Double(jobs.count)
        }

        endAppNapPrevention()
        releasePowerAssertion()
        isProcessing = false
    }

    // MARK: - Single Job Execution

    private func executeJob(_ job: CompressJob) async {
        guard let ffmpegURL = FFmpegLocator.ffmpegURL,
              FileManager.default.isExecutableFile(atPath: ffmpegURL.path) else {
            job.state = .failed
            job.errorMessage = "FFmpeg binary not found in app bundle."
            return
        }

        guard let outputURL = job.outputURL else {
            job.state = .failed
            job.errorMessage = "Output URL could not be determined."
            return
        }

        // Detect HDR via ffprobe color_transfer
        let colorTransfer = await DurationProbe.probeColorTransfer(url: job.file.url)
        let hdrMode: HDRMode = (colorTransfer?.contains("smpte2084") == true) ? .hdr10 : .sdr
        job.hdrMode = hdrMode

        let arguments = CompressArgumentBuilder.build(
            inputURL: job.file.url,
            outputURL: outputURL,
            configuration: job.configuration,
            hdrMode: hdrMode
        )

        let totalDuration = job.file.durationSeconds ?? 0

        job.state = .running
        job.progress = 0.0
        job.startDate = Date()
        job.appendLog("$ ffmpeg \(arguments.joined(separator: " "))")
        job.appendLog("HDR: \(hdrMode.displayName)")

        let capturedFFmpegURL = ffmpegURL
        let capturedArguments = arguments
        let capturedEnvironment = FFmpegLocator.processEnvironment()
        let capturedDuration = totalDuration
        let throttleMs = Self.uiUpdateIntervalMs

        await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
            nonisolated(unsafe) var didResume = false
            func safeResume() {
                guard !didResume else { return }
                didResume = true
                continuation.resume()
            }

            Task.detached(priority: .userInitiated) { [weak self] in
                let process = Process()
                process.executableURL = capturedFFmpegURL
                process.arguments = capturedArguments
                process.environment = capturedEnvironment

                let stderrPipe = Pipe()
                let stdoutPipe = Pipe()
                process.standardError = stderrPipe
                process.standardOutput = stdoutPipe

                let stderrHandle = stderrPipe.fileHandleForReading
                let stdoutHandle = stdoutPipe.fileHandleForReading

                func cleanupPipes() {
                    stderrHandle.readabilityHandler = nil
                    stdoutHandle.readabilityHandler = nil
                    try? stderrHandle.close()
                    try? stdoutHandle.close()
                }

                await MainActor.run { [weak self] in
                    self?.currentProcess = process
                    job.processHandle = process
                }

                // Throttled stderr parsing
                nonisolated(unsafe) var lastFlush = ContinuousClock.now
                nonisolated(unsafe) var pendFrame: Int?
                nonisolated(unsafe) var pendTime: String?
                nonisolated(unsafe) var pendProcessedSeconds: Double?
                nonisolated(unsafe) var pendProgress: Double?
                nonisolated(unsafe) var firstMetricWallDate: Date?
                nonisolated(unsafe) var firstMetricTimeSeconds: Double?

                stderrHandle.readabilityHandler = { handle in
                    let data = handle.availableData
                    guard !data.isEmpty,
                          let text = String(data: data, encoding: .utf8) else { return }

                    var logBatch: [String] = []

                    for line in text.split(separator: "\n", omittingEmptySubsequences: true) {
                        let lineStr = String(line)
                        if let prog = FFmpegProgress.parse(line: lineStr) {
                            pendFrame = prog.frame
                            pendTime = prog.time
                            pendProcessedSeconds = prog.timeSeconds
                            if capturedDuration > 0 {
                                pendProgress = min(prog.timeSeconds / capturedDuration, 1.0)
                            }
                        } else {
                            logBatch.append(lineStr)
                        }
                    }

                    let now = ContinuousClock.now
                    if (now - lastFlush) >= .milliseconds(throttleMs) || !logBatch.isEmpty {
                        let fr = pendFrame
                        let ti = pendTime
                        let ps = pendProcessedSeconds
                        let pr = pendProgress
                        let lg = logBatch
                        pendFrame = nil
                        pendTime = nil
                        pendProcessedSeconds = nil
                        pendProgress = nil
                        lastFlush = now

                        Task { @MainActor in
                            if let v = fr { job.currentFrame = v }
                            if let v = ti { job.currentTime = v }
                            if let processedSeconds = ps {
                                if firstMetricWallDate == nil && processedSeconds > 0 {
                                    firstMetricWallDate = Date()
                                    firstMetricTimeSeconds = processedSeconds
                                }

                                if let startWall = firstMetricWallDate,
                                   let startProcessed = firstMetricTimeSeconds {
                                    let elapsed = max(Date().timeIntervalSince(startWall), 0.001)
                                    let mediaDelta = max(processedSeconds - startProcessed, 0.0)

                                    if elapsed >= 0.35, mediaDelta > 0 {
                                        let speed = mediaDelta / elapsed
                                        let fps = Double(job.currentFrame) / elapsed
                                        job.speed = String(format: "x%.3f", speed)
                                        job.fps = String(format: "%.1f", fps)
                                    } else {
                                        job.speed = "warming..."
                                    }
                                }
                            }
                            if let v = pr { job.progress = v }
                            for l in lg { job.appendLog(l) }
                        }
                    }
                }

                stdoutHandle.readabilityHandler = { handle in
                    let data = handle.availableData
                    guard !data.isEmpty else { return }
                    // Compress uses -stats on stderr; stdout is unused but drain it
                }

                process.terminationHandler = { [weak self] proc in
                    cleanupPipes()

                    let fFrame = pendFrame
                    let fProg = pendProgress

                    Task { @MainActor [weak self] in
                        if let v = fFrame { job.currentFrame = v }
                        if let v = fProg { job.progress = v }

                        job.endDate = Date()
                        self?.currentProcess = nil
                        job.processHandle = nil

                        if proc.terminationStatus == 0 {
                            job.state = .completed
                            job.progress = 1.0
                            job.appendLog("✅ Compression completed successfully.")
                        } else if proc.terminationReason == .uncaughtSignal {
                            job.state = .cancelled
                            job.appendLog("🛑 Compression cancelled by user.")
                        } else {
                            job.state = .failed
                            job.errorMessage = "FFmpeg exited with code \(proc.terminationStatus)"
                            job.appendLog("❌ FFmpeg exited with code \(proc.terminationStatus)")
                        }

                        safeResume()
                    }
                }

                do {
                    try process.run()
                } catch {
                    cleanupPipes()
                    Task { @MainActor in
                        job.state = .failed
                        job.errorMessage = "Failed to launch FFmpeg: \(error.localizedDescription)"
                        job.endDate = Date()
                        job.appendLog("❌ Failed to launch: \(error.localizedDescription)")
                        safeResume()
                    }
                }
            }
        }
    }

    // MARK: - Power Assertion

    private func acquirePowerAssertion() {
        guard !hasPowerAssertion else { return }
        let reason = "Anime4K Upscaler — Compress is processing video" as CFString
        let result = IOPMAssertionCreateWithName(
            kIOPMAssertPreventUserIdleSystemSleep as CFString,
            IOPMAssertionLevel(kIOPMAssertionLevelOn),
            reason,
            &powerAssertionID
        )
        hasPowerAssertion = (result == kIOReturnSuccess)
    }

    private func releasePowerAssertion() {
        guard hasPowerAssertion else { return }
        IOPMAssertionRelease(powerAssertionID)
        hasPowerAssertion = false
        powerAssertionID = 0
    }

    private func beginAppNapPrevention() {
        guard activityToken == nil else { return }
        activityToken = ProcessInfo.processInfo.beginActivity(
            options: [.userInitiated, .idleSystemSleepDisabled],
            reason: "Video compression in progress"
        )
    }

    private func endAppNapPrevention() {
        if let token = activityToken {
            ProcessInfo.processInfo.endActivity(token)
            activityToken = nil
        }
    }
}

File: windows/src/viewmodels/CompressViewModel.cpp
#include "CompressViewModel.h"
#include "../utils/FFmpegLocator.h"
#include "../utils/DurationProbe.h"
#include "../utils/SleepPreventer.h"

#include <QFileInfo>
#include <QDir>
#include <QLocale>

CompressViewModel::CompressViewModel(QObject* parent) : QObject(parent) {
    m_throttleTimer = new QTimer(this);
    connect(m_throttleTimer, &QTimer::timeout, this, &CompressViewModel::handleThrottledUpdates);
}

CompressViewModel::~CompressViewModel() {
    cancelProcessing();
    qDeleteAll(m_jobs);
}

void CompressViewModel::setPickerService(IPickerService* picker) {
    m_picker = picker;
}

const QVector<VideoFile>& CompressViewModel::files() const {
    return m_files;
}

QUuid CompressViewModel::selectedFileID() const {
    return m_selectedFileID;
}

void CompressViewModel::setSelectedFileID(const QUuid& id) {
    if (m_selectedFileID != id) {
        m_selectedFileID = id;
        emit selectedFileChanged();
    }
}

VideoFile* CompressViewModel::selectedFile() {
    for (auto& f : m_files) {
        if (f.id == m_selectedFileID) return &f;
    }
    return nullptr;
}

void CompressViewModel::addFiles() {
    if (!m_picker) return;
    m_picker->pickFiles(
        "Select Video Files",
        "Video Files (*.mp4 *.mkv *.mov *.avi *.webm *.flv *.ts)",
        [this](QStringList paths) {
            if (!paths.isEmpty())
                addFilesFromDrop(paths);
        });
}

void CompressViewModel::addFilesFromDrop(const QStringList& paths) {
    for (const QString& path : paths) {
        QString clean = QDir::cleanPath(path);
        QFileInfo fi(clean);
        QString ext = fi.suffix().toLower();
        if (!supportedVideoExtensions().contains(ext)) continue;

        bool duplicate = false;
        for (const VideoFile& f : m_files) {
            if (f.filePath == clean) {
                duplicate = true;
                break;
            }
        }
        if (duplicate) continue;

        m_files.append(VideoFile::fromPath(clean));
    }

    probeNewFiles();
    if (m_selectedFileID.isNull() && !m_files.isEmpty()) {
        m_selectedFileID = m_files.first().id;
        emit selectedFileChanged();
    }
    emit filesChanged();
}

void CompressViewModel::removeFile(const QUuid& id) {
    for (int i = 0; i < m_files.size(); ++i) {
        if (m_files[i].id == id) {
            m_files.removeAt(i);
            break;
        }
    }
    if (m_selectedFileID == id) {
        m_selectedFileID = m_files.isEmpty() ? QUuid() : m_files.first().id;
        emit selectedFileChanged();
    }
    emit filesChanged();
}

void CompressViewModel::removeAllFiles() {
    m_files.clear();
    m_selectedFileID = QUuid();
    emit selectedFileChanged();
    emit filesChanged();
}

void CompressViewModel::probeNewFiles() {
    QStringList unprobed;
    for (const auto& file : m_files) {
        if (!file.durationSeconds.has_value()) {
            unprobed.append(file.filePath);
        }
    }
    if (unprobed.isEmpty()) return;

    DurationProbe::batchProbe(unprobed, [this](QMap<QString, std::optional<ProbeResult>> results) {
        for (auto it = results.begin(); it != results.end(); ++it) {
            QString path = it.key();
            auto res = it.value();
            if (res.has_value()) {
                for (auto& file : m_files) {
                    if (file.filePath == path) {
                        file.durationSeconds = res->durationSeconds;
                        file.width = res->width;
                        file.height = res->height;
                    }
                }
            }
        }
        emit filesChanged();
        emit selectedFileChanged();
    });
}

CompressEncoder CompressViewModel::encoder() const {
    return m_encoder;
}

void CompressViewModel::setEncoder(CompressEncoder enc) {
    if (m_encoder != enc) {
        m_encoder = enc;
        onEncoderChanged();
        emit configurationChanged();
    }
}

int CompressViewModel::quality() const {
    return m_quality;
}

void CompressViewModel::setQuality(int val) {
    if (m_quality != val) {
        m_quality = val;
        emit configurationChanged();
    }
}

ContentType CompressViewModel::contentType() const {
    return m_contentType;
}

void CompressViewModel::setContentType(ContentType type) {
    if (m_contentType != type) {
        m_contentType = type;
        emit configurationChanged();
    }
}

int CompressViewModel::bFrames() const {
    return m_bFrames;
}

void CompressViewModel::setBFrames(int val) {
    if (m_bFrames != val) {
        m_bFrames = val;
        emit configurationChanged();
    }
}

bool CompressViewModel::longGOPEnabled() const {
    return m_longGOPEnabled;
}

void CompressViewModel::setLongGOPEnabled(bool val) {
    if (m_longGOPEnabled != val) {
        m_longGOPEnabled = val;
        emit configurationChanged();
    }
}

void CompressViewModel::onEncoderChanged() {
    m_quality = defaultQuality(m_encoder);
}

void CompressViewModel::updateQuality(int value) {
    m_quality = qBound(0, value, maxQuality(m_encoder));
    emit configurationChanged();
}

QString CompressViewModel::outputDirectory() const {
    return m_outputDirectory;
}

void CompressViewModel::setOutputDirectory(const QString& path) {
    if (m_outputDirectory != path) {
        m_outputDirectory = QDir::cleanPath(path);
        emit outputDirectoryChanged();
    }
}

QString CompressViewModel::outputDirectoryDisplayName() const {
    if (m_outputDirectory.isEmpty()) return "Not selected";
    return QFileInfo(m_outputDirectory).fileName();
}

void CompressViewModel::selectOutputDirectory() {
    if (!m_picker) return;
    m_picker->pickDirectory(
        "Select Output Directory",
        m_outputDirectory,
        [this](QString dir) {
            if (!dir.isEmpty())
                setOutputDirectory(dir);
        });
}

CompressViewModel::ViewState CompressViewModel::viewState() const {
    return m_viewState;
}

bool CompressViewModel::isProcessing() const {
    return m_isProcessing;
}

int CompressViewModel::currentJobIndex() const {
    return m_currentJobIndex;
}

int CompressViewModel::totalJobs() const {
    return m_totalJobs;
}

double CompressViewModel::overallProgress() const {
    return m_overallProgress;
}

const QVector<CompressJob*>& CompressViewModel::jobs() const {
    return m_jobs;
}

bool CompressViewModel::canStartProcessing() const {
    return !m_files.isEmpty() && !m_isProcessing;
}

void CompressViewModel::startProcessing() {
    if (!canStartProcessing()) return;

    if (m_outputDirectory.isEmpty()) {
        selectOutputDirectory();
        if (m_outputDirectory.isEmpty()) return;
    }

    CompressConfiguration config;
    config.encoder = m_encoder;
    config.quality = m_quality;
    config.contentType = m_contentType;
    config.bFrames = m_bFrames;
    config.longGOPEnabled = m_longGOPEnabled;

    qDeleteAll(m_jobs);
    m_jobs.clear();

    for (const VideoFile& file : m_files) {
        m_jobs.append(new CompressJob(file, config, m_outputDirectory, this));
    }

    m_viewState = ViewState::Processing;
    emit viewStateChanged();

    executeBatch();
}

void CompressViewModel::cancelProcessing() {
    m_cancellationRequested = true;
    if (m_currentProcess && m_currentProcess->state() == QProcess::Running) {
        m_currentProcess->kill();
    }
}

void CompressViewModel::returnToConfiguration() {
    if (m_isProcessing) return;
    m_viewState = ViewState::Configuration;
    emit viewStateChanged();
}

QString CompressViewModel::totalFileSize() const {
    qint64 total = 0;
    for (const auto& f : m_files) {
        total += f.fileSizeBytes;
    }
    return QLocale().formattedDataSize(total, 2, QLocale::DataSizeSIFormat);
}

QString CompressViewModel::totalDuration() const {
    double total = 0.0;
    bool hasVal = false;
    for (const auto& file : m_files) {
        if (file.durationSeconds.has_value()) {
            total += file.durationSeconds.value();
            hasVal = true;
        }
    }
    if (!hasVal) return QString();
    int hours = static_cast<int>(total) / 3600;
    int minutes = (static_cast<int>(total) % 3600) / 60;
    int seconds = static_cast<int>(total) % 60;
    return QString("%1:%2:%3")
        .arg(hours, 2, 10, QChar('0'))
        .arg(minutes, 2, 10, QChar('0'))
        .arg(seconds, 2, 10, QChar('0'));
}

QString CompressViewModel::batchSummary() const {
    return QString("%1 file%2 \u2022 %3 \u2022 %4")
        .arg(m_files.size())
        .arg(m_files.size() == 1 ? "" : "s")
        .arg(displayName(m_encoder))
        .arg(displayName(m_contentType));
}

int CompressViewModel::completedJobCount() const {
    int count = 0;
    for (const auto* job : m_jobs) {
        if (job->state == JobState::Completed) count++;
    }
    return count;
}

int CompressViewModel::failedJobCount() const {
    int count = 0;
    for (const auto* job : m_jobs) {
        if (job->state == JobState::Failed) count++;
    }
    return count;
}

bool CompressViewModel::allJobsFinished() const {
    if (m_jobs.isEmpty()) return false;
    for (const auto* job : m_jobs) {
        if (!isTerminal(job->state)) return false;
    }
    return true;
}

void CompressViewModel::executeBatch() {
    if (m_jobs.isEmpty()) return;

    m_isProcessing = true;
    m_totalJobs = m_jobs.size();
    m_currentJobIndex = 0;
    m_overallProgress = 0.0;
    m_cancellationRequested = false;

    SleepPreventer::preventSleep();
    emit viewStateChanged();

    executeNext();
}

void CompressViewModel::executeNext() {
    if (m_cancellationRequested || m_currentJobIndex >= m_jobs.size()) {
        SleepPreventer::allowSleep();
        m_isProcessing = false;
        m_currentJob = nullptr;
        m_currentProcess = nullptr;
        emit viewStateChanged();
        return;
    }

    m_currentJob = m_jobs[m_currentJobIndex];
    m_currentJobIndex++;

    executeJob(m_currentJob);
}

void CompressViewModel::executeJob(CompressJob* job) {
    DurationProbe::probeColorTransfer(job->file.filePath, [this, job](std::optional<QString> colorTransfer) {
        HDRMode hdr = (colorTransfer.has_value() && colorTransfer->contains("smpte2084")) ? HDRMode::HDR10 : HDRMode::SDR;
        job->hdrMode = hdr;

        QString ffmpeg = FFmpegLocator::ffmpegPath();
        if (!FFmpegLocator::isFFmpegExecutable()) {
            job->state = JobState::Failed;
            job->errorMessage = "FFmpeg binary not found or not executable.";
            job->endDate = QDateTime::currentDateTime();
            job->appendLog("❌ Error: " + job->errorMessage);
            emit jobProgressUpdated(job);
            executeNext();
            return;
        }

        QStringList arguments = CompressArgumentBuilder::build(
            job->file.filePath,
            job->outputPath,
            job->configuration,
            hdr
        );

        job->state = JobState::Running;
        job->progress = 0.0;
        job->startDate = QDateTime::currentDateTime();
        job->appendLog("$ ffmpeg " + arguments.join(" "));
        job->appendLog("HDR Detection: " + displayName(hdr));
        emit jobProgressUpdated(job);

        m_firstMetricWallDate = QDateTime();
        m_firstMetricTimeSeconds = -1.0;

        m_currentProcess = new QProcess(this);
        m_currentProcess->setProgram(ffmpeg);
        m_currentProcess->setArguments(arguments);
        m_currentProcess->setProcessEnvironment(FFmpegLocator::processEnvironment());
        job->processHandle = m_currentProcess;

        connect(m_currentProcess, &QProcess::readyReadStandardError, this, &CompressViewModel::onStderrReady);
        connect(m_currentProcess, QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished),
                this, &CompressViewModel::onProcessFinished);

        m_currentProcess->start();
        m_throttleTimer->start(100);
    });
}

void CompressViewModel::onProcessFinished(int exitCode, QProcess::ExitStatus exitStatus) {
    m_throttleTimer->stop();
    handleThrottledUpdates();

    if (m_currentJob) {
        m_currentJob->endDate = QDateTime::currentDateTime();
        m_currentJob->processHandle = nullptr;

        if (exitCode == 0 && exitStatus == QProcess::NormalExit) {
            m_currentJob->state = JobState::Completed;
            m_currentJob->progress = 1.0;
            m_currentJob->appendLog("✅ Compression completed successfully.");
        } else if (m_cancellationRequested || exitStatus == QProcess::CrashExit) {
            if (m_cancellationRequested) {
                m_currentJob->state = JobState::Cancelled;
                m_currentJob->appendLog("🛑 Compression cancelled by user.");
            } else {
                m_currentJob->state = JobState::Failed;
                m_currentJob->errorMessage = QString("FFmpeg crashed or exited with code %1").arg(exitCode);
                m_currentJob->appendLog("❌ Error: " + m_currentJob->errorMessage);
            }
        } else {
            m_currentJob->state = JobState::Failed;
            m_currentJob->errorMessage = QString("FFmpeg exited with code %1").arg(exitCode);
            m_currentJob->appendLog("❌ Error: " + m_currentJob->errorMessage);
        }

        emit jobProgressUpdated(m_currentJob);
    }

    if (m_currentProcess) {
        m_currentProcess->deleteLater();
        m_currentProcess = nullptr;
    }

    m_overallProgress = static_cast<double>(m_currentJobIndex) / static_cast<double>(m_totalJobs);
    executeNext();
}

void CompressViewModel::onStderrReady() {
    if (!m_currentProcess || !m_currentJob) return;

    QByteArray data = m_currentProcess->readAllStandardError();
    QString text = QString::fromUtf8(data);

    QStringList lines = text.split('\n', Qt::SkipEmptyParts);
    for (const QString& line : lines) {
        QString trimmed = line.trimmed();
        auto progOpt = FFmpegProgress::parse(trimmed);
        if (progOpt.has_value()) {
            FFmpegProgress prog = progOpt.value();
            m_pendFrame = prog.frame;
            m_pendTime = prog.time;
            m_pendFps = QString::number(prog.fps, 'f', 1);

            double mediaSeconds = prog.timeSeconds();
            m_pendProgress = mediaSeconds; // Temp hijack progress for timeSeconds

            double duration = m_currentJob->file.durationSeconds.value_or(0.0);
            if (duration > 0.0) {
                m_pendProgress = qMin(mediaSeconds / duration, 1.0);
            }
        } else {
            m_logBatch.append(trimmed);
        }
    }
}

void CompressViewModel::handleThrottledUpdates() {
    if (!m_currentJob) return;

    bool updated = false;

    if (m_pendFrame.has_value()) {
        m_currentJob->currentFrame = m_pendFrame.value();
        m_pendFrame = std::nullopt;
        updated = true;
    }
    if (m_pendTime.has_value()) {
        m_currentJob->currentTime = m_pendTime.value();
        m_pendTime = std::nullopt;
        updated = true;
    }
    if (m_pendFps.has_value()) {
        m_currentJob->fps = m_pendFps.value();
        m_pendFps = std::nullopt;
        updated = true;
    }
    if (m_pendProgress.has_value()) {
        double progVal = m_pendProgress.value();
        m_currentJob->progress = progVal;

        // Retrieve the current media time in seconds
        double duration = m_currentJob->file.durationSeconds.value_or(0.0);
        double mediaSeconds = progVal * duration;

        if (mediaSeconds > 0.0) {
            if (!m_firstMetricWallDate.isValid()) {
                m_firstMetricWallDate = QDateTime::currentDateTime();
                m_firstMetricTimeSeconds = mediaSeconds;
            } else {
                double elapsed = qMax(m_firstMetricWallDate.msecsTo(QDateTime::currentDateTime()) / 1000.0, 0.001);
                double mediaDelta = qMax(mediaSeconds - m_firstMetricTimeSeconds, 0.0);

                if (elapsed >= 0.35 && mediaDelta > 0.0) {
                    double speed = mediaDelta / elapsed;
                    m_currentJob->speed = QString("x%1").arg(speed, 0, 'f', 3);
                } else {
                    m_currentJob->speed = "warming...";
                }
            }
        }

        m_pendProgress = std::nullopt;
        updated = true;
    }

    if (!m_logBatch.isEmpty()) {
        for (const QString& log : m_logBatch) {
            m_currentJob->appendLog(log);
        }
        m_logBatch.clear();
        updated = true;
    }

    if (updated) {
        emit jobProgressUpdated(m_currentJob);
    }
}
